README.md
PhlyBlog: Static Blog Generator
===============================

This module is a tool for generating a static blog.

Blog posts are simply PHP files that create and return `PhlyBlog\EntryEntity`
objects. You point the compiler at a directory, and it creates a tree of files
representing your blog and its feeds. These can either be consumed by your
application, or they can be plain old HTML markup files that you serve
directly.

Requirements
------------

- PHP >= `7.3`
- Laminas packages, notably: 
  - `Laminas\View\View`, used to render and write generated files
  - `Laminas\Mvc` and `Laminas\ModuleManager`, as this implements a module, and the
    compiler script depends on it and an `Application` instance. As such, it also
    has a dependency on `Laminas\ServiceManager` and `Laminas\EventManager`.
  - `Laminas\Feed\Writer`
  - `Laminas\Tag\Cloud`
- `PhlyCommon` (for Entity and Filter interfaces)

Installation
------------

You can add this module as a git submodule to your repository. Alternately, use
composer. To do so, add the following `composer.json` configuration in your
project:

```js
{
    "require": {
        "phly/phly-blog": "^3.0"
    }
}
```

and then execute:

```bash
php composer.phar install
```

Writing Entries
---------------

Find a location in your repository for entries, preferably outside your document
root; I recommend either `data/blog/` or `posts/`.

Post files are simply PHP files that return a `PhlyBlog\EntryEntity` instance.
A sample is provided in `misc/sample-post.php`. This post can be copied as a
template.

Important things to note:

- Set the created and/or updated timestamps. Alternately, use `DateTime` or
  `date()` to generate a timestamp based on a date/time string.
- Entries marked as "drafts" (i.e., `setDraft(true)`) will not be published.
- Entries marked as private (i.e., `setPublic(false)`) will be published, but
  will not be aggregated in paginated views or feeds. As such, you need to hand
  the URL to somebody in order for them to see it.
- You can set an array of tags. Tags can have whitespace, which will be
  translated to "+" characters.

Usage
=====

This module is ZF2 Console-aware. Once installed and active in your
application, you should be able to run:

```bash
% php public/index.php
```

and see usage for the module. Currently, it defines a "blog compile" action
that can generate the following artifacts:

- A file per entry
- Paginated entry files
- Paginated entry files by year
- Paginated entry files by month
- Paginated entry files by day
- Paginated entry files by tag
- Atom and/or RSS feeds for recent entries
- Atom and/or RSS feeds for recent entries by tag
- Optionally, a tag cloud

You will want to setup local configuration; I recommend putting it in
`config/autoload/module.blog.config.global.php`. As a sample:

    ```php
    <?php
    return array(
    'blog' => array(
        'options' => array(
            // The following indicate where to write files. Note that this
            // configuration writes to the "public/" directory, which would
            // create a blog made from static files. For the various
            // paginated views, "%d" is the current page number; "%s" is
            // typically a date string (see below for more information) or tag.
            'by_day_filename_template'   => 'public/blog/day/%s-p%d.html',
            'by_month_filename_template' => 'public/blog/month/%s-p%d.html',
            'by_tag_filename_template'   => 'public/blog/tag/%s-p%d.html',
            'by_year_filename_template'  => 'public/blog/year/%s-p%d.html',
            'entries_filename_template'  => 'public/blog-p%d.html',

            // In this case, the "%s" is the entry ID.
            'entry_filename_template'    => 'public/blog/%s.html',

            // For feeds, the final "%s" is the feed type -- "atom" or "rss". In
            // the case of the tag feed, the initial "%s" is the current tag.
            'feed_filename'              => 'public/blog-%s.xml',
            'tag_feed_filename_template' => 'public/blog/tag/%s-%s.xml',
             
            // This is the link to a blog post
            'entry_link_template'        => '/blog/%s.html',

            // These are the various URL templates for "paginated" views. The
            // "%d" in each is the current page number.
            'entries_url_template'       => '/blog-p%d.html',
            // For the year/month/day paginated views, "%s" is a string
            // representing the date. By default, this will be "YYYY",
            // "YYYY/MM", and "YYYY/MM/DD", respectively.
            'by_year_url_template'       => '/blog/year/%s-p%d.html',
            'by_month_url_template'      => '/blog/month/%s-p%d.html',
            'by_day_url_template'        => '/blog/day/%s-p%d.html',

            // These are the primary templates you will use -- the first is for
            // paginated lists of entries, the second for individual entries.
            // There are of course more templates, but these are the only ones 
            // that will be directly referenced and rendered by the compiler.
            'entries_template'           => 'phly-blog/list',
            'entry_template'             => 'phly-blog/entry',

            // The feed author information is default information to use when
            // the author of a post is unknown, or is not an AuthorEntity
            // object (and hence does not contain this information).
            'feed_author_email'          => 'you@your.tld',
            'feed_author_name'           => "Your Name Here",
            'feed_author_uri'            => 'http://your.tld',
            'feed_hostname'              => 'http://your.tld',
            'feed_title'                 => 'Blog Entries :: Your Blog Name',
            'tag_feed_title_template'    => 'Tag: %s :: Your Blog Name',

            // If generating a tag cloud, you can specify options for
            // Laminas\Tag\Cloud. The following sets up percentage sizing from
            // 80-300%
            'tag_cloud_options'          => array('tagDecorator' => array(
                'decorator' => 'html_tag',
                'options'   => array(
                    'fontSizeUnit' => '%',
                    'minFontSize'  => 80,
                    'maxFontSize'  => 300,
                ),
            )),
        ),
        
        // This is the location where you are keeping your post files (the PHP
        // files returning `PhlyBlog\EntryEntity` objects).
        'posts_path'     => 'data/posts/',

        // You can provide your own callback to setup renderer and response
        // strategies. This is useful, for instance, for injecting your 
        // rendered contents into a layout.
        // The callback will receive a View instance, application configuration
        // (as an array), and the application's Locator instance.
        'view_callback'  => array('Application\Module', 'prepareCompilerView'),

        // Tag cloud generation is possible, but you likely need to capture
        // the rendered cloud to inject elsewhere. You can do this with a
        // callback.
        // The callback will receive a Laminas\Tag\Cloud instance, the View
        // instance, application configuration // (as an array), and the
        // application's Locator instance.
        'cloud_callback' => array('Application\Module', 'handleTagCloud'),
    ),
    'di' => array('instance' => array(
        // You will likely want to customize the templates provided. Do so by
        // creating your own in your own module, and make sure you alter the
        // resolvers so that they point to the override locations. Below, I'm
        // putting my overrides in my Application module.
        'Laminas\View\Resolver\TemplateMapResolver' => array('parameters' => array(
            'map' => array(
                'phly-blog/entry-short'  => 'module/Application/view/phly-blog/entry-short.phtml',
                'phly-blog/entry'        => 'module/Application/view/phly-blog/entry.phtml',
                'phly-blog/list'         => 'module/Application/view/phly-blog/list.phtml',
                'phly-blog/paginator'    => 'module/Application/view/phly-blog/paginator.phtml',
                'phly-blog/tags'         => 'module/Application/view/phly-blog/tags.phtml',
            ),
        )),

        'Laminas\View\Resolver\TemplatePathStack' => array('parameters' => array(
            'paths' => array(
                'phly-blog' => 'module/Application/view',
            ),
        )),
    ));
    ```

When you run the command line tool, it will generate files in the locations you
specify in your configuration.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 2.1.2 - TBD

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 2.1.1 - 2020-11-09

### Changed

- [#31](https://github.com/phly/PhlyBlog/pull/31) changes the `Compiler` to implement `EventsCapableInterface` instead of `EventManagerAwareInterface` (the latter is a superset of the former). Doing so ensures that any initializers for `EventManagerAwareInterface` do not trigger, which prevents double-injection of the `EventManager` instance, and thus prevents overwriting any listeners attached via delegator factories. The `setEventManager()` method is still defined.


-----

### Release Notes for [2.1.1](https://github.com/phly/PhlyBlog/milestone/7)

2.1.x bugfix release (patch)

### 2.1.1

- Total issues resolved: **1**
- Total pull requests resolved: **1**
- Total contributors: **2**

#### bug

 - [31: Make Compiler EventsCapable instead of EventManagerAware](https://github.com/phly/PhlyBlog/pull/31) thanks to @weierophinney and @vrkansagara

## 2.1.0 - 2020-11-04

### Added

- [#29](https://github.com/phly/PhlyBlog/pull/29) adds a `PhlyBlog\CompilerFactory`, and wires the `PhlyBlog\Compiler` service to be created via that factory.

### Changed

- [#29](https://github.com/phly/PhlyBlog/pull/29) adds an optional `?Compiler $compiler = null` argument to the `CompileCommand` constructor. When provided, the command will use that `Compiler` instance. The `CompileCommandFactory` now pulls the `Compiler` service from the container and passes it for that argument.


-----

### Release Notes for [2.1.0](https://github.com/phly/PhlyBlog/milestone/3)



### 2.1.0

- Total issues resolved: **1**
- Total pull requests resolved: **1**
- Total contributors: **2**

#### enhancement

 - [29: Add Compiler as a container service](https://github.com/phly/PhlyBlog/pull/29) thanks to @weierophinney and @vrkansagara

## 2.0.2 - 2020-10-13

### Fixed

- [#26](https://github.com/phly/PhlyBlog/pull/26) fixes issues with display of progress bars, ensuring they display along with the label detailing what is being compiled.

- [#26](https://github.com/phly/PhlyBlog/pull/26) fixes the remaining issues with rendering the blog posts using the `phly-blog:compile` tooling, ensuring that the view is capable of identifying a renderer, posts are rendered within the layout, and that no duplication of metadata occurs.

-----

### Release Notes for [2.0.2](https://github.com/phly/PhlyBlog/milestone/5)

- Total issues resolved: **0**
- Total pull requests resolved: **1**
- Total contributors: **1**

#### bug

 - [26: Fix rendering of posts and pagination](https://github.com/phly/PhlyBlog/pull/26) thanks to @weierophinney

## 2.0.1 - 2020-10-13

### Fixed

- [#23](https://github.com/phly/PhlyBlog/pull/23) fixes an issue in the phly-blog:compile command with resolution of the Tags listener, which led to a fatal error.

-----

### Release Notes for [2.0.1](https://github.com/phly/PhlyBlog/milestone/2)

- Total issues resolved: **1**
- Total pull requests resolved: **1**
- Total contributors: **2**

#### bug

 - [23: Fix tags listener resolution](https://github.com/phly/PhlyBlog/pull/23) thanks to @weierophinney and @vrkansagara



src/ConfigProvider.php
<?php

namespace PhlyBlog;

class ConfigProvider
{
    public function __invoke(): array
    {
        return [
            'blog'            => $this->getBlogConfig(),
            'laminas-cli'     => $this->getCliConfig(),
            'router'          => [
                'routes' => [
                    'phly-blog' => $this->getRoutes(),
                ],
            ],
            'service_manager' => $this->getDependencyConfig(),
            'view_manager'    => $this->getViewConfig(),
        ];
    }

    public function getBlogConfig(): array
    {
        return [
            'options'        => [
                'author_feed_filename_template' => 'public/blog/author/%s-%s.xml',
                'author_feed_title_template'    => 'Author: %s',
                'by_author_filename_template'   => 'public/blog/author/%s-p%d.html',
                'by_day_filename_template'      => 'public/blog/day/%s-p%d.html',
                'by_month_filename_template'    => 'public/blog/month/%s-p%d.html',
                'by_tag_filename_template'      => 'public/blog/tag/%s-p%d.html',
                'by_year_filename_template'     => 'public/blog/year/%s-p%d.html',
                'entries_filename_template'     => 'public/blog-p%d.html',
                'entries_template'              => 'phly-blog/list',
                'entry_filename_template'       => 'public/blog/%s.html',
                'entry_link_template'           => '/blog/%s.html',
                'entry_template'                => 'phly-blog/entry',
                'feed_author_email'             => 'you@your.tld',
                'feed_author_name'              => "Your name here",
                'feed_author_uri'               => 'http://your.tld',
                'feed_filename'                 => 'public/blog-%s.xml',
                'feed_hostname'                 => 'http://your.tld',
                'feed_title'                    => 'Blog Entries',
                'tag_feed_filename_template'    => 'public/blog/tag/%s-%s.xml',
                'tag_feed_title_template'       => 'Tag: %s',
                'tag_cloud_options'             => ['tagDecorator' => [
                    'decorator' => 'html_tag',
                    'options'   => [
                        'fontSizeUnit' => '%',
                        'minFontSize'  => 80,
                        'maxFontSize'  => 300,
                    ],
                ]],
            ],
            'posts_path'     => 'data/blog/',
            'cloud_callback' => false,
        ];
    }

    public function getCliConfig(): array
    {
        return [
            'commands' => [
                'phly-blog:compile' => Console\CompileCommand::class,
            ],
        ];
    }

    public function getRoutes(): array
    {
        return [
            'type'          => 'Literal',
            'options'       => [
                'route' => '/blog',
            ],
            'may_terminate' => false,
            'child_routes'  => [
                'index'     => [
                    'type'    => 'Literal',
                    'options' => [
                        'route' => '.html',
                    ],
                ],
                'feed-atom' => [
                    'type'    => 'Literal',
                    'options' => [
                        'route' => '-atom.xml',
                    ],
                ],
                'feed-rss'  => [
                    'type'    => 'Literal',
                    'options' => [
                        'route' => '-rss.xml',
                    ],
                ],
                'entry'     => [
                    'type'    => 'Regex',
                    'options' => [
                        'regex' => '/(?<id>[^/]+)\.html',
                        'spec'  => '/%id%.html',
                    ],
                ],
                'author'    => [
                    'type'          => 'Regex',
                    'options'       => [
                        'regex'    => '/author/(?<author>[^/]+)',
                        'defaults' => [
                            'action' => 'author',
                        ],
                        'spec'     => '/author/%author%',
                    ],
                    'may_terminate' => false,
                    'child_routes'  => [
                        'page'      => [
                            'type'    => 'Literal',
                            'options' => [
                                'route' => '.html',
                            ],
                        ],
                        'feed-atom' => [
                            'type'    => 'Literal',
                            'options' => [
                                'route' => '-atom.xml',
                            ],
                        ],
                        'feed-rss'  => [
                            'type'    => 'Literal',
                            'options' => [
                                'route' => '-rss.xml',
                            ],
                        ],
                    ],
                ],
                'tag'       => [
                    'type'          => 'Regex',
                    'options'       => [
                        'regex'    => '/tag/(?<tag>[^/.-]+)',
                        'defaults' => [
                            'action' => 'tag',
                        ],
                        'spec'     => '/tag/%tag%',
                    ],
                    'may_terminate' => false,
                    'child_routes'  => [
                        'page'      => [
                            'type'    => 'Literal',
                            'options' => [
                                'route' => '.html',
                            ],
                        ],
                        'feed-atom' => [
                            'type'    => 'Literal',
                            'options' => [
                                'route' => '-atom.xml',
                            ],
                        ],
                        'feed-rss'  => [
                            'type'    => 'Literal',
                            'options' => [
                                'route' => '-rss.xml',
                            ],
                        ],
                    ],
                ],
                'year'      => [
                    'type'    => 'Segment',
                    'options' => [
                        'route'       => '/year/:year.html',
                        'constraints' => [
                            'year' => '\d{4}',
                        ],
                        'defaults'    => [
                            'action' => 'year',
                        ],
                    ],
                ],
                'month'     => [
                    'type'    => 'Segment',
                    'options' => [
                        'route'       => '/month/:year/:month.html',
                        'constraints' => [
                            'year'  => '\d{4}',
                            'month' => '\d{2}',
                        ],
                        'defaults'    => [
                            'action' => 'month',
                        ],
                    ],
                ],
                'day'       => [
                    'type'    => 'Segment',
                    'options' => [
                        'route'       => '/day/:year/:month/:day.html',
                        'constraints' => [
                            'year'  => '\d{4}',
                            'month' => '\d{2}',
                            'day'   => '\d{2}',
                        ],
                        'defaults'    => [
                            'action' => 'day',
                        ],
                    ],
                ],
            ],
        ];
    }

    public function getDependencyConfig(): array
    {
        return [
            'factories' => [
                Console\CompileCommand::class => Console\CompileCommandFactory::class,
                Console\View::class           => Console\ViewFactory::class,
            ],
        ];
    }

    public function getViewConfig(): array
    {
        return [
            'template_map'        => [
                'phly-blog/entry-short' => __DIR__ . '/../view/phly-blog/entry-short.phtml',
                'phly-blog/entry'       => __DIR__ . '/../view/phly-blog/entry.phtml',
                'phly-blog/list'        => __DIR__ . '/../view/phly-blog/list.phtml',
                'phly-blog/paginator'   => __DIR__ . '/../view/phly-blog/paginator.phtml',
                'phly-blog/tags'        => __DIR__ . '/../view/phly-blog/tags.phtml',
            ],
            'template_path_stack' => [
                'phly-blog' => __DIR__ . '/../view',
            ],
        ];
    }
}
